B2g dev allow unverified

shane-tomlinson opened the pull request:
@jedp - allow unverified!

seanmonstar committed:
fixing cef_logger.alert, and INSERT INTO user
(cherry-picked from 98f9e06930531b6199f7edd42870c5eba4eec994)

ozten committed:
I think a recent merge removed these forceIssuer pieces
(Cherry-picked commit 56624d2320235b5418f18856b807c1f4de52f79f)

seanmonstar committed:
email inserted in mysql should be `unverified`
(Cherry-picked commit 270aad40591bfb659ab24ab4dfa4db0f73a86462)

ozten committed:
The verifier service should be explicit about unverified email in its output
(cherry-picked commit a81a55b9d8e323f6fcad25ecd966d1d8735526fb)

seanmonstar committed:
authenticate with unverified email and allowUnverified

fixes #2902

(Cherry-picked commit 3c32c43e0ef7c0a4543237ac320b088f47f22cf2)
Conflicts:
	resources/static/dialog/js/modules/authenticate.js

(cherry-picked commit 438a80e3e3c91d6ee05628de5dc07cbde87c3941)
Conflicts:
	resources/static/dialog/js/misc/state.js
	resources/static/dialog/js/modules/authenticate.js

seanmonstar committed:
send a confirmation email instead of newUser email

since a user is already being created, newUser is the wrong email to
send

(Cherry-picked commit 06f94c3e2f695a9c6274ff4f52f631537a54d198)

seanmonstar committed:
remove unverified certs once an email is verified

fixes #2924

(Cherry-picked commit d308c0d176521124a8cac754279e10a6cb915859)

(Cherry-picked commit 97f242f64924c96e55c8908405cafab1c7959dfa)
Conflicts:
	resources/static/common/js/user.js

seanmonstar committed:
mark password reseted emails as verified

if an email is not verified, and the user goes through the
reset password flow, then the email really should be verified.

fixes #2925

(Cherry-picked commit 06721c3483d0c474aa4af6b5c2bad5d5714a4738)

Conflicts:
	lib/db/mysql.js

seanmonstar committed:
removing duplicate emailInfo (wut)
(Cherry-picked commit 73cd47c1ca93cf94a93e06dd00f285fd20aad33e)

seanmonstar committed:
tests for verifying email after password reset

(Cherry-picked commit 649302b4bc8ea8e9fa9612d7ff2f77eb6c63f579)

Conflicts:
	tests/unverified-email-test.js

seanmonstar committed:
make JSON db delete old staged rows
(Cherry-picked commit b9a64140c3c6cbb4fbe99c44f611ee2e8d8dd895)

jedp committed:
An implementation for B2G to ask allow users to log in even if
Persona has not verified that the user owns the email address.

Starts with navigator.id.request({ allowUnverified: true });
When the assertion is returned to the RP, and if they are using
the Persona verifier, must be sure to include an allowUnverified=true
parameter to the verifier, otherwise it will be rejected.

Behind the scenes, when the dialog tries to stage the user, instead
the user is created immediately, an email is added in our email table,
and a stagedEmail token is created and send to the user. When
allowUnverified is true, our wsapi will also check for a staged email/pass,
and use that to allow the user to login.

(merged from commit ebe21f3e3fc9bb80f65f694b359e377ddfad64a0)
(cherry-picked commit bc4ac912df1e8899c954f75ece6727f1309acab6)

Conflicts:
	example/rp/index.html
	resources/static/common/js/network.js
	resources/static/dialog/js/misc/helpers.js
	resources/static/dialog/js/misc/state.js
	resources/static/dialog/js/modules/dialog.js

jedp committed:
unverified email now adds passwd to user row

this makes things a whole lot simpler underneath

(merged from commit 427052a8f908bc02344f021c85e60c19675fef6e)

jedp committed:
added function call accidentally dropped in commit c0707ddf

jedp committed:
Fixing errors introduced in merge in commit c0707ddf

Conflicts:
	resources/static/pages/js/forgot.js

jedp committed:
Removed password param in reset password test (see commit 3c04102e)

Commit:
Move the allowUnverified code from network.js to user.js

Update tests. Add a test for network.js->certKey

(Cherry-picked commit f783c700416316692d1f3457fdde32d903e78b9b)

Conflicts:
	resources/static/common/js/network.js
	resources/static/common/js/user.js
	resources/static/test/cases/common/js/network.js

Commit:
Change allowUnverified to _experimental_allowUnverified when calling request.

Commit:
Move addressInfo cache update for unverified users from dialog/js/helpers.js to user.js

Commit:
Get the tests passing again (back & front)!

cleanup.

jedp commented:
OMG

jedp commented on lib/db/json.js:
```diff
@@ -199,13 +201,14 @@ exports.emailToUID = function(email, cb) {
+  var owned = m && m.length === 1;
   process.nextTick(function() {
-    cb(null, m && m.length === 1);
+    cb(null, owned);
```
thank you - helpful clarification

jedp commented on lib/keysigner/ca.js:
```diff
@@ -25,11 +25,17 @@ function parsePublicKey(serializedPK) {
+  //TODO: use a different secret if unverified
+
+  var principal = {};
+  principal[ unverified ? 'unverified-email' : 'email' ] = email;
```
Some time in the near future (let's say after the merge is all done), I think we should rename `unverified-email` to `unverified_email`, so it's easier to access as a property of a javascript object.

shane-tomlinson replied:
I agree, as long as it follows existing conventions.

jedp commented on resources/static/common/js/user.js:
```diff
@@ -265,20 +268,31 @@ BrowserID.User = (function() {
-    if (onComplete) onComplete(true);
+    // XXX This needs to be looked at to make sure caching does not bite us.
+    // I would rather pass in the unverified flag so that caching does not
+    // cause problems here.
```
Good catch of what might be a subtle bug.  Let's file a follow-up issue to investigate after the merge?

jedp commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -228,10 +244,18 @@ BrowserID.State = (function() {
       }
+      else if(self.newFxAccountEmail) {
+        startAction(false, "doStageUser", info);
+// TODO         startAction(false, "doStageResetPassword", info); ???
```
is this still a TODO?

shane-tomlinson replied:
Dunno, I made the same comment here: https://github.com/mozilla/browserid/pull/3240#discussion_r3895102

jedp commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -122,6 +122,12 @@ BrowserID.Modules.Dialog = (function() {
   }
 
+  function fixupIssuer(url) {
+    // XXX do something awesome here.
```
Shall we file a follow-up issue?

shane-tomlinson replied:
It's already taken care of in the b2g_dev_force_issuer branch. Rebasing this and pushing.

jedp commented:
@ozten and @seanmonstar, this looks good to me.  But it's such a huge PR I'd be grateful to have another pair of eyes go over it before pushing the big green button.  Can one of you take a second look?

jedp commented:
Don't merge until forceIssuer PR lands!

Commit:
Remove the extra pubKey.serialize() that is blowing things up.

(Cherry-picked commit d13b5d1ef57b9011b03e8166c3f0bdbd2c3c9ecf)

Conflicts:
	resources/static/common/js/user.js

shane-tomlinson commented:
@jedp, @seanmonstar - this should have a significantly smaller diff once the force-issuer branch has landed. 

In other news, this makes me happy (the instance is called force-issuer, but the code is allow-unverified):

```
➜  automation-tests git:(b2g_dev_allow_unverified) ✗ ./scripts/run-all.js -e force-issuer -p 5
adding add-primary-to-primary.js
adding add-primary-to-secondary.js
adding cancel-account.js
adding change-password-test.js
adding frontend-qunit-test.js
adding health-check-tests.js
adding broken-primary.js
adding primary-shuts-down-single-email.js
adding primary-starts-up-single-email.js
adding new-user-primary-test.js
adding new-user-secondary-test.js
ignoring public-terminals.js
adding remove-email.js
adding reset-password-test.js
adding returning-user.js
adding sign-in-test.js
Running 15 suite(s) on 1 platform(s), 5 at a time against https://force-issuer.personatest.org
>>>>>.........................<>.......<>.....<>.........................<>.......<>.................<>..........<>....................................<>......<>.......<>...........................<...............<......<.....<...<

216/216 tests passed
➜  automation-tests git:(b2g_dev_allow_unverified) ✗
```

shane-tomlinson commented:
when adding an email with "unverified allowed", I am currently forced to verify the email.

shane-tomlinson commented on bin/verifier:
```diff
@@ -135,13 +147,11 @@ function doVerification(req, resp, next) {
-        email : r.success.email,
         audience : audience, // NOTE: we return the audience formatted as the RP provided it, not normalized in any way.
-        expires : new Date(r.success.expires).valueOf(),
-        issuer: r.success.issuer
```
We could use a comment here about what comes back from the verifier. It took me a minute to grok why we removed email - I assume it is because the verifier could return either "email" or "unverified-email"

shane-tomlinson commented:
@jedp - It looks like there are many cases not handled by this PR.

If "allowUnverified" is set to true, I think the following flows need updated as well:
- add email
- transition_no_password
- transition_to_secondary

Also, do we still send primary users down the primary flow if allowUnverified is set to true? It makes sense to allow this.

seanmonstar commented:
@shane-tomlinson yes, this doesn't prevent primaries from authenticating, but if they do, they may require verification before giving back an assertion. We can't enforce allowUnverified on them. Which is why, to ensure a user doesn't need to verify, it must be used with forceIssuer as well.

Commit:
Fix the tests for the updates to the force-issuer branch.

shane-tomlinson commented on lib/keysigner/ca.js:
```diff
@@ -25,11 +25,17 @@ function parsePublicKey(serializedPK) {
     return cb("expiration cannot be null");
 
-  cert.sign({publicKey: publicKey, principal: {email: email}},
+  //TODO: use a different secret if unverified
```
Do we have to add a different key here?

jedp replied:
I think this is a question we need @benadida to weigh in on

benadida replied:
longer term, we probably want a different issuer, but short term, this is okay.

shane-tomlinson commented on bin/verifier:
```diff
@@ -93,6 +96,7 @@ function doVerification(req, resp, next) {
   } else if (req.body && req.body.forceIssuer) {
     forceIssuer =  req.body.forceIssuer;
   }
+  var allowUnverified = (req.query && req.query.allowUnverified) || req.body.allowUnverified;
```
Are we allowing allowUnverified as a GET parameter? If not, the req.query should be taken off.

shane-tomlinson replied:
Do we allow GET parameters elsewhere?

jedp replied:
It seems we do allow GET params, yes

shane-tomlinson commented on lib/db/json.js:
```diff
@@ -476,6 +473,7 @@ exports.addPrimaryEmailToAccount = function(userID, emailToAdd, cb) {
 };
 
 exports.checkAuth = function(userID, cb) {
+  logger.debug("checkAuthh for userID " + userID);
```
Is this still needed?

shane-tomlinson commented on lib/db/mysql.js:
```diff
@@ -492,9 +492,14 @@ exports.completePasswordReset = function(secret, password, cb) {
         function(err) {
           if (err) return cb(err);
 
-          // update the password!
```
In the JSON adapter, deleteOldSecrets has been added several times, but the same treatment has not been added here. Is this expected or does a deleteOldSecrets equivalent need to be added here?

jedp replied:
Good catch.  Hm.  I'm not sure.  It looks like the json database has `staged` and `stagedUser`, whereas the mysql has just `staged`, which it overwrites both on `stageUser` and `stageEmail`.  Looks like the json database is getting the same result by deleting both rows when either is updated?

jedp replied:
Ja.  @seanmonstar confirms that this is ok.

shane-tomlinson commented on lib/wsapi.js:
```diff
@@ -22,6 +22,7 @@ config = require('./configuration'),
 logger = require('./logging.js').logger,
 httputils = require('./httputils.js'),
 forward = require('./http_forward.js').forward,
+booleanQuery = require('./boolean-query'),
```
Do any wsapi calls need booleanQuery? booleanQuery only works with GET parameters.

jedp replied:
Good catch. looks like we can remove this.

shane-tomlinson commented on lib/wsapi/complete_reset.js:
```diff
@@ -70,7 +70,7 @@ exports.process = function(req, res) {
         });
       } else {
-        return httputils.authRequired(res, 'Provide your password');
+        return httputils.authRequired(res, 'Provide your password- params were: ' + JSON.stringify(req.params) );
```
Do we need the extra `- params were...`?

jedp replied:
kill the debugging cruft

shane-tomlinson commented on lib/wsapi/complete_reset.js:
```diff
@@ -88,7 +88,7 @@ exports.process = function(req, res) {
         });
       } else {
-        return httputils.authRequired(res, 'Provide your password');
+        return httputils.authRequired(res, 'Provide your password' + JSON.stringify(req.params));
```
Likewise.

shane-tomlinson commented on lib/wsapi/stage_user.js:
```diff
@@ -58,30 +64,51 @@ exports.process = function(req, res) {
+              return wsapi.databaseDown(res, err);
+            }
+
+            if (!req.session) req.session = {};
```
From here to the else statement is duplicated below and should be collapsed. (the only difference is the unverified: true in res.json)

jedp replied:
They have the same shape, but they're actually pretty different, I think: Different `db` functions, callback signatures, cef logging, `req.session` variables, and send-email functions.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -489,8 +510,18 @@ BrowserID.User = (function() {
+        network.createUser.bind(network, email, password,
+            origin, allowUnverified), function(status) {
+              // update the cache if necessary
+              // XXX this needs a test.
```
Add the test.

shane-tomlinson commented on resources/static/dialog/js/misc/helpers.js:
```diff
@@ -79,13 +79,20 @@
-        var info = { email: email, password: password };
-        self.publish("user_staged", info, info);
+        var msg = { email: email, password: password };
+        // XXX these need tests.
```
Add the test.

shane-tomlinson commented on resources/static/dialog/js/misc/helpers.js:
```diff
@@ -79,13 +79,20 @@
+        var msg = { email: email, password: password };
+        // XXX these need tests.
+        if (status.unverified) {
+          msg.type = "secondary";
```
Both of these "msg" settings are implicit in the message "unverified_created"

shane-tomlinson replied:
I'm not convinced these are actually needed. Are they used somewhere?

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -423,12 +431,18 @@ BrowserID.State = (function() {
       }
       else {
+        // make sure an unverified-certs are removed
+        if (record.unverified && info.state !== 'unverified') {
```
This seems like it should be taken care of in user.js

shane-tomlinson commented on resources/static/dialog/js/modules/actions.js:
```diff
@@ -95,6 +95,13 @@ BrowserID.Modules.Actions = (function() {
+    doAuthenticateWithUnverifiedEmail: function(info) {
+      var self = this;
+      dialogHelpers.authenticateUser.call(this, info.email, info.password, function() {
+        self.publish("authenticated", info);
```
From https://github.com/mozilla/browserid/pull/3307/files#r3992382 above (implicit), set the info.type and info.unverified here. here.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -274,6 +274,8 @@ BrowserID.Modules.Dialog = (function() {
         // forceAuthentication is used by the Marketplace to ensure that the
         // user knows the password to this account. We ignore any active session.
         if (paramsFromRP._experimental_forceAuthentication &&
+            // TODO this and _experimental_allowUnverified should throw an
```
Make a better check like the TODO says.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -285,6 +287,14 @@ BrowserID.Modules.Dialog = (function() {
+        // verified their email address in order to send an assertion.
+        // if the user *has* verified, it will be a verified assertion.
+        if (paramsFromRP._experimental_allowUnverified &&
+            true === paramsFromRP._experimental_allowUnverified) {
```
This needs a better check as well.

shane-tomlinson commented on resources/static/test/cases/dialog/js/modules/dialog.js:
```diff
@@ -542,6 +542,15 @@
     );
   });
 
+  // TODO Add a test for forceIssuer
```
Add a forceIssuer test if it doesn't already exist.

Commit:
Add a validateBoolean, use this for forceAuthentication and allowUnverified.

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -423,12 +431,18 @@ BrowserID.State = (function() {
       else {
+        // make sure an unverified-certs are removed
+        if (record.unverified && info.state !== 'unverified') {
+          storage.invalidateEmail(email);
```
This only invalidates emails for the default issuer, should this invalidate emails for the currently forced issuer?

Commit:
Add tests for helpers.js->createUser with allowUnverified and user.js->createSecondaryUser with allowUnverified.

shane-tomlinson commented:
@jedp, @seanmonstar, @benadida - what should happen, if for the same issuer, the user visits one RP and set up an account with an unverified email address and they visit another RP where allowUnverified is not set? Currently, there is a discrepancy if the user is authenticated to Persona vs if they are not.

If they _are_ signed in, they are forced to verify their email and sent to the verification screen.
If they _are not_ signed in, they are sent down the normal new user flow, including setting the account password.

This somewhat makes sense. In the case of a "stolen" account, the real owner of the address could come along and take the email address back from the original user.

This caused me to test another case where there is a verified problem:

User A, who is not the real owner of the address, visits a site that allows unverified emails
User B, who is the real owner of the address, visits the same RP.
User B is asked for the password User A typed. User B does not know the password. User B cannot sign in even though they own the address.

Even if User B opens the email address that was sent when User A registered the address, User B will still not know the password and cannot claim their own email address for themselves. 

wah wah.

Commit:
Add tests for users who type in a known unverified address. Fix the problem of a user typing one of these addresses just sitting there.

shane-tomlinson commented:
So my last comment, the rightful owner of the address can still use password reset. What should happen if allowUnverified is set to true, should User B be forced to validate their email before they can snag it off of User A, or should they be signed straight in?

jedp committed:
code cleanup; no new logic

seanmonstar commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -159,6 +159,13 @@ BrowserID.Modules.Dialog = (function() {
 
+  function validateBoolean(bool, name) {
+    if (typeof bool !== "boolean") {
+      throw new Error("invalid value for " + name + ": " + bool);
```
a tangential conversation: should we be throwing these errors in the dialog, or right when they try to do it with `navigator.id.request({ _experimental_forceIssuer: "heck yes" });`

shane-tomlinson replied:
@seanmonstar - we throw some errors in the dialog, some from include.js. The advantage to the dialog, for request based parameters, is that checks only exist in one place. If we pushed the checks to a lower level, we would have to check in include.js as well as internal_api.js. I do think there are advantages to seeing the errors early though, as console messages - "why doesn't the browser pop up? oh, there's an error in the console."

shane-tomlinson commented on resources/static/common/js/network.js:
```diff
@@ -15,6 +15,7 @@ BrowserID.Network = (function() {
       xhr = bid.XHR,
       post = xhr.post,
       get = xhr.get,
+      // XXX get this out of here!
```
storage was not added for this PR, this is just a note to get it out.

shane-tomlinson commented:
@jedp - what more do we have to do to merge this PR?

Commit:
When invalidating certs, invalidate certs for the current issuer.

jedp commented:
Oh frabjous day.
